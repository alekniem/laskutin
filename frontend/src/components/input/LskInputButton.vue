<template>
    <v-btn :color="color" :disabled="disabled" :id="cid" depressed small>
        {{ buttonText | trans }}
    </v-btn>
</template>

<script>
export default {
    name: "LskInputButton",

    props: {
        buttonText: { type: String, required: true },
        cid: { type: String, required: true },
        color: { type: String, required: false, default: "primary" },
        disabled: { type: Boolean, required: false, default: false },
    },
};
</script>
