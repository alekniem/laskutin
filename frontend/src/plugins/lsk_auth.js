import lsk_api from "@/plugins/lsk_api";

export default {
    async login(username, password, cb) {
        if (localStorage.getItem("token")) {
            if (cb) cb(true, "");
            this.onChange(true);
            return;
        }

        try {
            await lsk_api.getCsrfCookie();
        } catch (error) {
            if (cb) cb(false, "Error in backend system");
            this.onChange(false);
            return;
        }

        try {
            await lsk_api.login(username, password);
            localStorage.setItem("token", "1");
            if (cb) cb(true, "");
            this.onChange(true);
        } catch (error) {
            if (cb) cb(false, "Failed to login");
            this.onChange(false);
        }
    },

    async logout(cb) {
        try {
            await lsk_api.logout();
            localStorage.removeItem("token");
            if (cb) cb(false);
            this.onChange(false);
        } catch (error) {
            if (cb) cb(true);
            this.onChange(true);
        }
    },

    loggedIn() {
        return !!localStorage.getItem("token");
    },

    onChange() {},
};
