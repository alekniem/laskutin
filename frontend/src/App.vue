<template>
    <v-app>
        <v-app-bar app color="primary" dark flat>
            <v-toolbar-title>
                <LskLogo />
            </v-toolbar-title>
            <v-spacer />
            <LskInputButton
                @click.native="processLogout"
                buttonText="Logout"
                cid="logoutButton"
                color="blue darken-4"
                v-if="loggedIn"
            />
            <template #extension>
                <v-tabs v-if="loggedIn">
                    <v-tab :to="{ name: 'indexInvoice' }" id="invoiceTab">
                        {{ "Invoices" | trans }}
                    </v-tab>
                    <v-tab :to="{ name: 'indexCustomer' }" id="customerTab">
                        {{ "Customers" | trans }}
                    </v-tab>
                    <v-tab :to="{ name: 'indexBiller' }" id="billerTab">
                        {{ "Billers" | trans }}
                    </v-tab>
                </v-tabs>
                <v-tabs v-else>
                    <v-tab :to="{ name: 'login' }" id="loginTab">
                        {{ "Login" | trans }}
                    </v-tab>
                </v-tabs>
            </template>
        </v-app-bar>
        <v-main>
            <div class="ma-5">
                <router-view />
            </div>
        </v-main>
    </v-app>
</template>

<script>
import lsk_auth from "@/plugins/lsk_auth";

export default {
    name: "App",

    data() {
        return {
            loggedIn: lsk_auth.loggedIn(),
        };
    },

    created() {
        lsk_auth.onChange = (loggedIn) => {
            this.loggedIn = loggedIn;
        };
    },

    methods: {
        async processLogout() {
            await lsk_auth.logout((loggedIn) => {
                if (!loggedIn) {
                    this.$router.replace({ name: "login" });
                }
            });
        },
    },
};
</script>
