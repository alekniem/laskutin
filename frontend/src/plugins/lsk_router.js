import VueRouter from "vue-router";
import lsk_auth from "@/plugins/lsk_auth";
import lsk_routes from "@/plugins/lsk_routes";

export function beforeEach(to, from, next) {
    if (lsk_auth.loggedIn()) {
        if (to.name === "index" || to.name === "login") {
            next({ name: "indexInvoice" });
        } else {
            next();
        }
    } else {
        if (to.name === "login") {
            next();
        } else {
            next({ name: "login" });
        }
    }
}

const router = new VueRouter({
    routes: lsk_routes,
    mode: "history",
});

router.beforeEach((to, from, next) => beforeEach(to, from, next));

export default router;
