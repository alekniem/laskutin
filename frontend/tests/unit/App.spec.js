import flushPromises from "flush-promises";
import App from "@/App";
import LskInputButton from "@/components/input/LskInputButton";
import lsk_auth from "@/plugins/lsk_auth";
import lsk_router from "@/plugins/lsk_router";
import { createWrapper } from "UT/lsk_test_helper";

const stubs = {
    LskInputButton,
    LskLogo: true,
};

jest.mock("@/components/page/LskPageLogin.vue", () => ({
    render: (h) => h("div", "LOGIN PAGE"),
}));

jest.mock("@/components/page/LskPageIndexInvoice.vue", () => ({
    render: (h) => h("div", "INVOICE INDEX PAGE"),
}));

jest.mock("@/plugins/lsk_auth.js", () => ({
    login: jest.fn(),
    logout: jest.fn(),
    loggedIn: jest.fn(),
}));

let fakeLoggedIn = false;

lsk_auth.login.mockImplementation(() => {
    fakeLoggedIn = true;
    lsk_auth.onChange(true);
});

lsk_auth.logout.mockImplementation(() => {
    fakeLoggedIn = false;
    lsk_auth.onChange(false);
});

lsk_auth.loggedIn.mockImplementation(() => fakeLoggedIn);

describe("App", () => {
    it("should match snapshot", async () => {
        const wrapper = createWrapper(App, {
            router: lsk_router,
            stubs,
        });
        await flushPromises();

        expect(wrapper.vm._data.loggedIn).toBe(false);
        expect(wrapper.html()).toMatchSnapshot("login page");

        lsk_auth.login();
        lsk_router.replace({ name: "indexInvoice" });
        await flushPromises();

        expect(wrapper.vm._data.loggedIn).toBe(true);
        expect(wrapper.html()).toMatchSnapshot("invoice index page");

        wrapper.find("#logoutButton").trigger("click");
        await flushPromises();

        expect(wrapper.vm._data.loggedIn).toBe(false);
        expect(lsk_auth.logout.mock.calls.length).toBe(1);
    });
});
