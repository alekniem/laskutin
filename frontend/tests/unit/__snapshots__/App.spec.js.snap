// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot: invoice index page 1`] = `
<div data-app="true"
     class="v-application v-application--is-ltr theme--light"
     id="app"
>
  <div class="v-application--wrap">
    <header class="v-sheet theme--dark v-toolbar v-toolbar--extended v-toolbar--flat v-app-bar v-app-bar--fixed primary"
            style="height: 112px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;"
            data-booted="true"
    >
      <div class="v-toolbar__content"
           style="height: 64px;"
      >
        <div class="v-toolbar__title">
          <lsklogo-stub>
          </lsklogo-stub>
        </div>
        <div class="spacer">
        </div>
        <button type="button"
                class="v-btn v-btn--has-bg theme--dark v-size--small blue darken-4"
                id="logoutButton"
        >
          <span class="v-btn__content">
            Logout
          </span>
        </button>
      </div>
      <div class="v-toolbar__extension"
           style="height: 48px;"
      >
        <div class="v-tabs theme--dark">
          <div role="tablist"
               class="v-item-group theme--dark v-slide-group v-tabs-bar v-tabs-bar--is-mobile white--text"
               data-booted="true"
          >
            <div class="v-slide-group__prev v-slide-group__prev--disabled">
              <transition-stub name="fade-transition">
              </transition-stub>
            </div>
            <div class="v-slide-group__wrapper">
              <div class="v-slide-group__content v-tabs-bar__content">
                <div class="v-tabs-slider-wrapper"
                     style="height: 2px; left: 0px; width: 0px;"
                >
                  <div class="v-tabs-slider">
                  </div>
                </div>
                <a href="/invoice"
                   aria-current="page"
                   class="v-tab--active v-tab"
                   tabindex="0"
                   aria-selected="true"
                   role="tab"
                   id="invoiceTab"
                >
                  Invoices
                </a>
                <a href="/customer"
                   class="v-tab"
                   tabindex="0"
                   aria-selected="false"
                   role="tab"
                   id="customerTab"
                >
                  Customers
                </a>
                <a href="/biller"
                   class="v-tab"
                   tabindex="0"
                   aria-selected="false"
                   role="tab"
                   id="billerTab"
                >
                  Billers
                </a>
              </div>
            </div>
            <div class="v-slide-group__next v-slide-group__next--disabled">
              <transition-stub name="fade-transition">
              </transition-stub>
            </div>
          </div>
        </div>
      </div>
    </header>
    <main class="v-main"
          style="padding: 112px 0px 0px 0px;"
          data-booted="true"
    >
      <div class="v-main__wrap">
        <div class="ma-5">
          <div>
            INVOICE INDEX PAGE
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;

exports[`App should match snapshot: login page 1`] = `
<div data-app="true"
     class="v-application v-application--is-ltr theme--light"
     id="app"
>
  <div class="v-application--wrap">
    <header class="v-sheet theme--dark v-toolbar v-toolbar--extended v-toolbar--flat v-app-bar v-app-bar--fixed primary"
            style="height: 112px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;"
    >
      <div class="v-toolbar__content"
           style="height: 64px;"
      >
        <div class="v-toolbar__title">
          <lsklogo-stub>
          </lsklogo-stub>
        </div>
        <div class="spacer">
        </div>
      </div>
      <div class="v-toolbar__extension"
           style="height: 48px;"
      >
        <div class="v-tabs theme--dark">
          <div role="tablist"
               class="v-item-group theme--dark v-slide-group v-tabs-bar v-tabs-bar--is-mobile white--text"
          >
            <div class="v-slide-group__prev v-slide-group__prev--disabled">
              <transition-stub name="fade-transition">
              </transition-stub>
            </div>
            <div class="v-slide-group__wrapper">
              <div class="v-slide-group__content v-tabs-bar__content">
                <div class="v-tabs-slider-wrapper"
                     style="transition: none;"
                >
                  <div class="v-tabs-slider">
                  </div>
                </div>
                <a href="/login"
                   aria-current="page"
                   class="v-tab--active v-tab"
                   tabindex="0"
                   aria-selected="true"
                   role="tab"
                   id="loginTab"
                >
                  Login
                </a>
              </div>
            </div>
            <div class="v-slide-group__next v-slide-group__next--disabled">
              <transition-stub name="fade-transition">
              </transition-stub>
            </div>
          </div>
        </div>
      </div>
    </header>
    <main class="v-main"
          style="padding: 112px 0px 0px 0px;"
    >
      <div class="v-main__wrap">
        <div class="ma-5">
          <div>
            LOGIN PAGE
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;
